app/src/main/java/org/mozilla/focus/shortcut/IconGenerator.kt

```diff
@@ -36,7 +37,15 @@ class IconGenerator {
          * on top of a generic launcher icon shape that we provide.
          */
         @JvmStatic
-        fun generateLauncherIcon(context: Context, url: String?): Bitmap {
+        fun generateLauncherIcon(context: Context, url: String?) = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            generateAdaptiveLauncherIcon(context, url)
+        } else {
+            // It'd more natural to return a Bitmap from generateLauncherIconPreOreo but we want to
+            // match the type of generateAdaptiveLauncherIcon so we wrap it in a Drawable.
+            BitmapDrawable(context.resources, generateLauncherIconPreOreo(context, url))
```

Comment:
But in AddToHomescreenDialogFragment we convert it back to a Bitmap to pass it to HomeScreen? I wonder:
* If we convert back to Bitmap - why not return Bitmap everywhere?
* I'm confused that we convert to bitmap at all - is the adaptive icon still "adaptive" after that?

Reply:
> If we convert back to Bitmap - why not return Bitmap everywhere?

Converting to Bitmap is a loss of information from an AdaptiveIconDrawable so I felt it was more correct to return the drawable (and the BitmapDrawable for the legacy icon just wraps the bitmap so it's fine).

> I'm confused that we convert to bitmap at all - is the adaptive icon still "adaptive" after that?

No, but I could not find any way of creating dynamic homescreen shortcuts while preserving AdaptiveIconDrawable, e.g. how [`createWithAdaptiveBitmap` will just pass null as the background drawable](http://androidxref.com/8.0.0_r4/xref/frameworks/base/graphics/java/android/graphics/drawable/Icon.java#294). That being said, in the AdaptiveIconPlayground test app, the homescreen shortcuts do not appear inside of it so I wonder if the layer-animating parts of Adaptive Icons are not available for the home screen and are only available when you display the application list.

tbh, the APIs seemed unintuitive and underdocumented so I wonder if they just didn't think through this API well enough. :\